src/handler/mod.rs

```diff
@@ -102,15 +109,90 @@ pub enum HandlerIn {
+pub enum RoutingType {
+    /// A session has been established with a peer.
     ///
     /// A session is only considered established once we have received a signed ENR from the
-    /// node and either the observed `SocketAddr` matches the one declared in the ENR or the
+    /// peer and either the observed `SocketAddr` matches the one declared in the ENR or the
     /// ENR declares no `SocketAddr`.
-    Established(Enr, SocketAddr, ConnectionDirection),
+    NoNatOrPortForward(Enr, SocketAddr, ConnectionDirection),
+    /// A session has been established with a peer behind an asymmetric NAT, either via the peer
+    /// simply dialing out to this node, via the NAT traversal protocol.
+    ///
+    /// A session with a peer behind an asymmetric NAT is only considered established once we have
+    /// received an ENR from the peer and the observed `SocketAddr` matches the one declared in
+    /// the 'nat'/'nat6' and 'udp'/'udp6' fields of the ENR.
+    ///
+    /// The [`ConnectionDirection`] is incoming if the peer dialed out to us directly without use
+    /// of the NAT traversal protocol. This peer will send PINGs to this node more frequently than
+    /// nodes that are not behind a NAT do, this way the hole punched in the peer's NAT for this
+    /// node will stay open and this peer can listen for requests from this node.
+    ///
+    /// Using the NAT traversal protocol to get the peer to dial out to us will always result in a
+    /// [`ConnectionDirection::Outgoing`] direction for both parties, the initiator and the
+    /// receiver. The initiator learns about the receiver in a NODES response from the rendezvous
+    /// node and attempts to connect to it via the rendezvous node. The receiver learns about the
+    /// initiator in a RELAYREQUEST from the rendezvous node and dials out to the initiator (this
+    /// first message is dropped by the initiator if it is also behind a NAT).
+    AsymmetricNat(Enr, SocketAddr, ConnectionDirection),
+    /// A session has been established with a peer behind a symmetric NAT.
+    ///
+    /// A session with a peer behind a symmetric NAT is only considered established once we have
+    /// received an ENR from the node and the observed `IpAddr` matches the one declared in the
+    /// 'nat'/'nat6' field of the ENR.
+    ///
+    /// The [`ConnectionDirection`] is always [`ConnectionDirection::Incoming`] as peers behind
+    /// symmetric NATs only dial out. This peer will send PINGs to this node more frequently than
+    /// nodes that are not behind a NAT do, this way the hole punched in the peer's NAT for this
+    /// node will stay open and this peer can listen for requests from this node.
+    ///
+    /// Peers behind symmetric NATs cannot be hole-punched as they cannot not advertise a port for
+    /// new peers to hole-punch them on. The [`ConnectionDirection`] is always incoming as peers
+    /// behind symmetric NATs only dial out.
+    SymmetricNat(Enr, SocketAddr),
+}
```

Comment:
No we don't because it has disappeared in some of @AgeManning commits!

The place where we really have to watch out is when doing the "am I behind a NAT?" voting and any variables that are set based on the peers correctly setting their ENR up till that point. Since if one incoming connection exists, and if this incoming connection is marked incoming based on the peers ENR, then we will believe we are behind a NAT.

Reply:
Oh no I found it, it has been divided up into NatKind and ConnectionState in a new file called `src/connection.rs`